### PR TITLE
fix(ai): apply input and output schemas when loading .prompt files

### LINKED ISSUE #299: Dotprompt: input/output schemas in .prompt files are not applied when loaded from promptDir
## What happens

When a prompt is loaded from a `.prompt` file via `promptDir`, the `input` and `output` schemas declared in the frontmatter are not applied. There are two separate problems.

1. The `input.schema` is dropped entirely. The loaded prompt action ends up with `inputSchema == null`. The Developer UI reads that schema to render the input form, so the prompt page shows no input fields, and input is never validated or parsed.

2. The `output.schema` is passed through without converting Picoschema to JSON Schema. The raw Picoschema (for example `{category: "string, the predicted category"}`) is sent to the model as the response schema, which is not valid JSON Schema.

Both schemas work correctly when the same prompt is defined in code with `definePrompt`. The problem is specific to the folder loader.

## Repro

`prompts/classify.prompt`:

```
---
model: googleai/gemini-flash-latest
input:
  schema:
    text: string, the text to classify
output:
  format: json
  schema:
    category: string, the predicted category
    confidence: number, a score from 0 to 1
---
{{role "user"}}
Classify: {{text}}
```

```dart
final ai = Genkit(plugins: [googleAI()], promptDir: './prompts');
final p = await ai.prompt('classify');
final opts = await p.render({'text': 'hello'});
print(opts.output?.toJson());
```

Output:

```
{format: json, jsonSchema: {category: string, the predicted category, confidence: number, a score from 0 to 1}}
```

The `jsonSchema` is the raw Picoschema. It should be:

```
{type: object, properties: {category: {type: string, description: ...}, confidence: {type: number, description: ...}}, required: [category, confidence]}
```

For the input problem, run the app under `genkit start` and request the action list from the reflection server (`GET /api/actions`), or open the prompt in the Developer UI. The `inputSchema` for the prompt action is `null` and the prompt page has no input fields, even though `input.schema` is declared in the file.

## Root cause

`loadPromptFolder` / `_loadPrompt` in `prompt_loader_io.dart` reads metadata from `dotpromptRegistry.parse(source)`. `parse` returns the raw frontmatter and does not resolve Picoschema. The Picoschema to JSON Schema conversion only happens in `_resolveMetadata`, which is reached through `render` / `renderMetadata`, not through `parse`.

So the loader:

- never reads `metadata.input`, and never sets `inputSchema` on the `PromptConfig` (and therefore on the registered `PromptAction`).
- copies `metadata.output.schema` straight into the output config without converting Picoschema.

One related detail worth flagging: `Picoschema.isPicoschema` does not recognize the `type, description` form (for example `name: string, the person to greet`) that the docs and examples use. It only matches bare type strings like `string`, `string[]`, and enums. A conversion guard based on `isPicoschema` would skip the most common case. `Picoschema.toJsonSchema` itself handles that form fine, so the fix should detect "already JSON Schema" directly rather than rely on `isPicoschema`.

## Impact

- The Developer UI shows no input form for any file-based prompt that declares `input.schema`. This is confusing since the schema is sitting right there in the file.
- Structured output from a file-based prompt with a Picoschema `output.schema` sends an invalid schema to the model. With the Google GenAI plugin this becomes `responseJsonSchema`, so the request can fail or the schema is ignored.

## Environment

- genkit 0.14.0
- genkit_google_genai 0.2.8
- dotprompt 0.0.1
- Dart 3.x

I have a fix ready and will open a PR linking back to this issue.

### PR DESCRIPTION
**Pull request by chrisraygill**
Fixes #299.

## Problem

When a prompt is loaded from a `.prompt` file via `promptDir`, the `input` and `output` schemas in the frontmatter are not applied. The folder loader reads metadata from `dotpromptRegistry.parse(source)`, which returns the raw frontmatter and does not resolve Picoschema (that only happens in `_resolveMetadata`, reached through `render` / `renderMetadata`).

As a result:

- `input.schema` was never set on the prompt action, so the action had a null input schema. The Developer UI renders its input form from that schema, so file-based prompts showed no input fields, and input was never validated.
- `output.schema` was copied through as raw Picoschema. That raw value was then sent to the model as the response schema. With the Google GenAI plugin it becomes `responseJsonSchema`, which is not valid JSON Schema, so the request can fail or the schema is ignored.

Both schemas already work when a prompt is defined in code with `definePrompt`. This was specific to the folder loader.

## Changes

In `prompt_loader_io.dart`:

- Resolve `input.schema` and set it on the `PromptConfig` (and therefore the registered `PromptAction`) as a `SchemanticType`.
- Convert `output.schema` to JSON Schema before building the output config.
- Both go through a small `_toJsonSchema` helper that converts Picoschema and leaves existing JSON Schema untouched.

The conversion does not gate on `Picoschema.isPicoschema`. That helper does not recognize the `type, description` form (for example `name: string, the person to greet`) used throughout the docs, so gating on it would skip the most common case. `_isJsonSchema` checks for an existing JSON Schema shape (a top-level standard `type`, or a `$schema` / `$ref` / `$defs` key) and everything else is treated as Picoschema and converted.

## Testing

Added three tests to the `loadPromptFolder` group in `prompt_test.dart`:

- input Picoschema (including an optional field) is wired onto the action and converted to JSON Schema with the right `required` set.
- plain JSON Schema for `input.schema` passes through unchanged.
- output Picoschema is converted to JSON Schema before it reaches the model.

Full `dart test` for the genkit package passes (368 tests). `dart analyze` is clean.

Before, for a prompt with `output.schema` of `{category: "string, the predicted category"}`:

```
{format: json, jsonSchema: {category: string, the predicted category}}
```

After:

```
{format: json, jsonSchema: {type: object, properties: {category: {type: string, description: the predicted category}}, ...}}
```

## Notes / follow-ups

- `Picoschema.isPicoschema` not matching the `type, description` form looks like a bug in the `dotprompt` package itself. This PR works around it in the loader. It may be worth fixing upstream too.
- Input default values from `input.default` are still not applied at render time. That is a separate render-path change and is out of scope here.

### CODE REVIEW

File: packages/genkit/lib/src/ai/prompt_loader_io.dart [hunk 1]
```diff
@@ -185,3 +198,59 @@ String _registryDefinitionKey(String name, String? variant, String? ns) {
...
+    'object',
+    'array',
+    'string',
+    'number',
+    'integer',
+    'boolean',
+    'null',
+  };
+  return jsonSchemaTypes.contains(schema['type']);
+}
```
[the remark this answers is not in the document]
Comment (chrisraygill) [reply packages/genkit/lib/src/ai/prompt_loader_io.dart:268]: Good catch. Updated `_isJsonSchema` in 7feb3f3 to also detect `properties`, `items`, `anyOf`, `oneOf`, and `allOf`, so a JSON Schema that omits the top-level `type` is no longer treated as Picoschema. Added a test for the no-top-level-type case.

File: packages/genkit/lib/src/ai/prompt_loader_io.dart [hunk 2]
```diff
@@ -185,3 +198,59 @@ String _registryDefinitionKey(String name, String? variant, String? ns) {
...
+SchemanticType<Map<String, dynamic>>? _toInputSchema(
+  Map<String, dynamic>? schema,
+) {
+  final jsonSchema = _toJsonSchema(schema);
+  if (jsonSchema == null) return null;
+  return SchemanticType.from<Map<String, dynamic>>(
+    jsonSchema: jsonSchema,
+    parse: (json) => (json as Map).cast<String, dynamic>(),
+  );
+}
```
[the remark this answers is not in the document]
Comment (chrisraygill) [reply packages/genkit/lib/src/ai/prompt_loader_io.dart:291]: Done in 7feb3f3. `parse` now guards the cast (`json is Map ? json.cast() : {}`). This path is reachable because a prompt can be invoked with null input and `Action.runRaw` still calls `inputSchema.parse(input)`. Added a test that parsing null returns normally.

**Comment by chrisraygill** [comment]
Addressed both review comments in 7feb3f3 and did another pass over the change.

Verified edge cases (added tests for the first two):
- JSON Schema that omits the top-level `type` and uses `properties` is kept as JSON Schema, not re-converted.
- A prompt invoked with null input parses to an empty map instead of throwing.
- `output: { format: json }` with no schema produces `{format: json}` and does not crash.
- A named schema reference (`input: schema: MyType`) passes through as `{$ref: MyType}` without crashing.

One known limitation, out of scope here: a named schema reference is passed through as a raw `$ref` and is not resolved against schemas registered with `defineSchema`, so the Developer UI cannot expand it into fields. Resolving named refs for file-loaded prompts would be a separate change. Input `default` values are also still not applied at render time.

**Review by pavelgj — APPROVED** [review]
